Mathlib/CategoryTheory/Sites/EffectiveEpimorphic.lean

```diff
@@ -0,0 +1,417 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.Sieves
+import Mathlib.CategoryTheory.Limits.Shapes.KernelPair
+
+/-!
+
+# Effective epimorphisms
+
+We define the notion of effective epimorphic (pre)sieves, morphisms and family of morphisms
+and provide some API for relating the three notions.
+
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable {C : Type _} [Category C]
+
+/-- A sieve is effective epimorphic if the associated cocone is a colimit cocone. -/
+def Sieve.EffectiveEpimorphic {X : C} (S : Sieve X) : Prop :=
+  Nonempty (IsColimit (S : Presieve X).cocone)
+
+/-- A presieve is effective epimorphic if the cocone assocaited to the sieve it generates
+is a colimit cocone. -/
+abbrev Presieve.EffectiveEpimorphic {X : C} (S : Presieve X) : Prop :=
+  (Sieve.generate S).EffectiveEpimorphic
+
+/--
+The sieve of morphisms which factor through a given morphism `f`.
+This is equal to `Sieve.generate (Presieve.singleton f)`, but has
+more convenient definitional properties.
+-/
+def Sieve.generateSingleton {X Y : C} (f : Y ⟶ X) : Sieve X where
+  arrows Z g := ∃ (e : Z ⟶ Y), e ≫ f = g
+  downward_closed := by
+    rintro W Z g ⟨e,rfl⟩ q
+    refine ⟨q ≫ e, by simp⟩
+
+lemma Sieve.generateSingleton_eq {X Y : C} (f : Y ⟶ X) :
+    Sieve.generate (Presieve.singleton f) = Sieve.generateSingleton f := by
+  ext Z ; intro g
+  constructor
+  · rintro ⟨W,i,p,⟨⟩,rfl⟩
+    exact ⟨i,rfl⟩
+  · rintro ⟨g,h⟩
+    exact ⟨Y,g,f,⟨⟩,h⟩
+
+/--
+This structure encodes the data required for a morphism to be an effective epimorphism.
+-/
+structure EffectiveEpiStruct {X Y : C} (f : Y ⟶ X) where
+  desc : ∀ {W : C} (e : Y ⟶ W),
+    (∀ {Z : C} (g₁ g₂ : Z ⟶ Y), g₁ ≫ f = g₂ ≫ f → g₁ ≫ e = g₂ ≫ e) → (X ⟶ W)
+  fac : ∀ {W : C} (e : Y ⟶ W)
+    (h : ∀ {Z : C} (g₁ g₂ : Z ⟶ Y), g₁ ≫ f = g₂ ≫ f → g₁ ≫ e = g₂ ≫ e),
+    f ≫ desc e h = e
+  uniq : ∀ {W : C} (e : Y ⟶ W)
+    (h : ∀ {Z : C} (g₁ g₂ : Z ⟶ Y), g₁ ≫ f = g₂ ≫ f → g₁ ≫ e = g₂ ≫ e)
+    (m : X ⟶ W), f ≫ m = e → m = desc e h
+
+attribute [nolint docBlame]
+  EffectiveEpiStruct.desc
+  EffectiveEpiStruct.fac
+  EffectiveEpiStruct.uniq
+
+/--
+A morphism `f : Y ⟶ X` is an effective epimorphism provided that `f` exhibits `X` as a colimit
+of the diagram of all "relations" `R ⇉ Y`.
+If `f` has a kernel pair, then this is equivalent to showing that the corresponding cofork is
+a colimit.
+-/
+class EffectiveEpi {X Y : C} (f : Y ⟶ X) : Prop where
+  effectiveEpi : Nonempty (EffectiveEpiStruct f)
+
+attribute [nolint docBlame] EffectiveEpi.effectiveEpi
+
+/-- Some chosen `EffectiveEpiStruct` associated to an effective epi. -/
+noncomputable
+def EffectiveEpi.struct {X Y : C} (f : Y ⟶ X) [EffectiveEpi f] :
```

Comment:
In the name, don't you want to make it apparent that this is not a canonical choice? Something like `EffectiveEpi.someStruct`?

Reply:
The name we use in the limits library is `getLimitCone`, so I can change this to `getStruct` for the sake of consistency.

Reply:
Changed to `getStruct`.